src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -2777,6 +2867,54 @@ inline bool VM_HeapWalkOperation::collect_stack_roots() {
   return true;
 }
 
+// Reports the unmounted virtual thread as JVMTI_HEAP_REFERENCE_THREAD,
+// walks the stack of the thread, finds all references (locals
+// and JNI calls) and reports these as stack references.
+inline bool VM_HeapWalkOperation::collect_vthread_stack_roots(oop vt) {
+  if (!JvmtiEnvBase::is_vthread_alive(vt)) {
+    return true;
+  }
+  ContinuationWrapper c(java_lang_VirtualThread::continuation(vt));
+  if (c.is_empty()) {
+    return true;
+  }
+  assert(!c.is_mounted(), "sanity check");
+
+  stackChunkOop chunk = c.last_nonempty_chunk();
+  if (chunk == nullptr || chunk->is_empty()) {
+    return true;
+  }
+
+  // vframes are resource allocated
+  Thread* current_thread = Thread::current();
+  ResourceMark rm(current_thread);
+  HandleMark hm(current_thread);
+
+  StackChunkFrameStream<ChunkFrames::Mixed> fs(chunk);
```

Comment:
There are ways to avoid using the `StackChunkFrameStream`.
You can find good examples in the jvmtiEnvBase.cpp.

Reply:
Fixed